### PR TITLE
Feature/graphical tests

### PR DESCRIPTION
**Pull request by Brennan1994**
Tests on graphical added to debug some viewmodel issues. Proves all input probs come back as outputs. and all input values come back as the means of the output distributions

### CODE REVIEW

File: StatisticsTests/GraphicalRelationships/GraphicalTests.cs [hunk 1]
```diff
@@ -51,30 +51,37 @@ public void GraphicalFunction_Test()
...
-        public void ReturnDistributionsForInputProbabilities()
+        [Theory]
+        [InlineData( new double[] {.002,.01,.02,.1,.5,.99},new double[] {146000,86000,66900,34900,2000,500},5,true,true  )] //Based on Elkhorn River at Highway 91 Dodge County FIS 2008
+        [InlineData(new double[] { .99, .95, .90, .85, .8, .75, .7, .65, .6, .55, .5, .45, .4, .35, .3, .25, .2, .15, .1, .05, .02, .01, .005, .0025 }, 
+            new double[] { 6.6, 7.4, 8.55, 9.95, 11.5, 12.7, 13.85, 14.7, 15.8, 16.7, 17.5, 18.25, 19, 19.7, 20.3, 21.1, 21.95, 23, 24.2, 25.7, 27.4, 28.4, 29.1, 29.4 },
+            20,true,true)]
+        public void ReturnDistributionsForInputProbabilities(double[] probs, double[] flows, int erl, bool usingFlows, bool flowsNotLogged)
         {
-            Graphical graphical = new Graphical(exceedanceProbabilities, quantileValues, equivalentRecordLength);
+            Graphical graphical = new Graphical(probs, flows, erl, usingFlows: true, flowsAreNotLogged: true);
```
Comment (HenryGeorgist) [inline StatisticsTests/GraphicalRelationships/GraphicalTests.cs]: shouldnt you use the bool flows are not logged that comes in from the inline data?

File: StatisticsTests/GraphicalRelationships/GraphicalTests.cs [hunk 2]
```diff
@@ -52,13 +52,13 @@ public void GraphicalFunction_Test()
...
-        [InlineData( new double[] {.002,.01,.02,.1,.5,.99},new double[] {146000,86000,66900,34900,2000,500},5,true,true  )] //Based on Elkhorn River at Highway 91 Dodge County FIS 2008
+        [InlineData(new double[] { .99, .5, .1, .02, .01, .002 }, new double[] { 500, 2000, 34900, 66900, 86000, 146000 }, 10, true, false)] //Based on Elkhorn River at Highway 91 Dodge County FIS 2008
         [InlineData(new double[] { .99, .95, .90, .85, .8, .75, .7, .65, .6, .55, .5, .45, .4, .35, .3, .25, .2, .15, .1, .05, .02, .01, .005, .0025 }, 
             new double[] { 6.6, 7.4, 8.55, 9.95, 11.5, 12.7, 13.85, 14.7, 15.8, 16.7, 17.5, 18.25, 19, 19.7, 20.3, 21.1, 21.95, 23, 24.2, 25.7, 27.4, 28.4, 29.1, 29.4 },
-            20,true,true)]
+            20,false,true)]
         public void ReturnDistributionsForInputProbabilities(double[] probs, double[] flows, int erl, bool usingFlows, bool flowsNotLogged)
         {
-            Graphical graphical = new Graphical(probs, flows, erl, usingFlows: true, flowsAreNotLogged: true);
+            Graphical graphical = new Graphical(probs, flows, erl, usingFlows: usingFlows, flowsAreNotLogged: flowsNotLogged);
```
Comment (HenryGeorgist) [inline StatisticsTests/GraphicalRelationships/GraphicalTests.cs:61]: good. disregard earlier comment

File: StatisticsTests/GraphicalRelationships/GraphicalTests.cs [hunk 3]
```diff
@@ -67,7 +67,7 @@ public void ReturnDistributionsForInputProbabilities(double[] probs, double[] fl
             }
         }
         [Theory]
-        [InlineData(new double[] { .99,.5,.1,.02,.01,.002 }, new double[] { 500,2000,34900,66900,86000,146000 }, 10, true, false)] //Based on Elkhorn River at Highway 91 Dodge County FIS 2008
+        [InlineData(new double[] { .99,.5,.1,.02,.01,.002 }, new double[] { 500,2000,34900,66900,86000,146000 }, 5, true, false)] //Based on Elkhorn River at Highway 91 Dodge County FIS 2008
```
Comment (HenryGeorgist) [inline StatisticsTests/GraphicalRelationships/GraphicalTests.cs:70]: 5 years is a really low eyrl

File: StatisticsTests/GraphicalRelationships/GraphicalTests.cs [hunk 4]
```diff
@@ -99,6 +99,43 @@ public void ReturnSameNumberOfProbabilitesAsDistributions(double[] probs, double
...
+            Graphical graphical = new Graphical(probs, flows, erl, usingFlows: usingFlows, flowsAreNotLogged: flowsNotLogged);
+            graphical.ComputeGraphicalConfidenceLimits();
+            Normal[] dists = graphical.FlowOrStageDistributions;
+            double[] means = new double[dists.Length];
+            double[] conf95 = new double[dists.Length];
+            double[] conf05 = new double[dists.Length];
+            for (int i = 0; i < dists.Length; i++)
+            {
+                means[i] = dists[i].Mean;
+                conf95[i] = dists[i].InverseCDF(.95);
```
Comment (HenryGeorgist) [inline StatisticsTests/GraphicalRelationships/GraphicalTests.cs:118]: might consider .99 and .01

**Review by HenryGeorgist — APPROVED** [review]

**Comment by rnugent3** [comment]
My brain is dizzy trying to solve a monotonicity problem. Given a non-monotonically increasing upper quantile, I can calculate how much larger the standard deviation needs to be so that the upper quantile is monotonically increasing. Likewise, given a non-monotonically increasing lower quantile, I can calculate how much smaller the standard deviation needs to be so that the lower quantile is monotonically increasing. 

But what if, by fixing one side, we mess up the other?

**Comment by rnugent3** [comment]
I think that the errors have come about because we're fixing one side and messing up the other. 

So we need to calculate the change in the standard deviation so that both lower and upper quantiles are monotonically increasing.

**Comment by rnugent3** [comment]
Aha. I need to reduce this to two equations and one unknown (or four equations and three unknowns). I think I found a path.

**Comment by rnugent3** [comment]
I arrive at the standard deviations being equal.

**Comment by rnugent3** [comment]
if the standard deviations are the same but the means are equal, then the quantiles equal each other plus the difference of the means where the means are assumed equal.

**Comment by rnugent3** [comment]
The current failing test is also a result of the monotonicity logic. Where standard deviation increases fast enough, the standard deviation of distribution i=1 is passed down to every other distribution i = 2, ... n. So in this case, setting the standard deviations equal does not work.

**Comment by rnugent3** [comment]
Bob had a separate SampleGraphical method. I disregarded that method because I wanted to create an uncertain paired data object with the graphical data and use SamplePairedData. 

It seems like we should instead just let stage_frequency be a graphical object when not created from an LP3 and rating curve and then use a Graphical.SampleGraphical method.

**Comment by HenryGeorgist** [comment]
@rnugent3 if you want to do that, that works - i think uncertain paired data implements an interface - IPairedDataProducer
you might just implement the graphical sample under that interface and we may be able to refer to it as the iPairedDtaProducer instaead of uncertainpaired data?

**Comment by rnugent3** [comment]
@Brennan1994 anything more you'd like to do here?

**Review by HenryGeorgist — APPROVED** [review]
im interested to se the sample method in fda model

**Pull request merged**